Practical Example/IosInterviewPracticeProject/IosInterviewPracticeProject/Screens/CourseListVC.swift

```diff
@@ -0,0 +1,147 @@
+//
+
+import UIKit
+
+class CourseListVC: UIViewController {
+  private var courseContentUrl: String?
+  private var courses = [Item]()
+  private var dataSource: UITableViewDiffableDataSource<Section, Item>?
+  private let courseManager = CourseAPIService()
+  private let timeFormatter: DateComponentsFormatter = {
+    let timeFormatter = DateComponentsFormatter()
+    timeFormatter.allowedUnits = [.hour, .minute]
+    timeFormatter.unitsStyle = .short
+    return timeFormatter
+  }()
+  @IBOutlet var courseList: UITableView!
+  @IBAction func filterCourses(_ sender: UISegmentedControl) {
+    var filteredContent = [Item]()
+
+    switch sender.selectedSegmentIndex {
+    case 0:
+      filteredContent = courses
+    case 1:
+      filteredContent = courses.filter { $0.attributes.contentType == ContentType.article }
+    case 2:
+      filteredContent = courses.filter { $0.attributes.contentType == ContentType.collection }
+    default:
+      filteredContent = courses
+    }
+
+    DispatchQueue.main.async {
+      self.createSnapshot(from: filteredContent.sorted { $0.attributes.releasedAt > $1.attributes.releasedAt })
+    }
+  }
+
+    override func viewDidLoad() {
+      super.viewDidLoad()
+      courseList.delegate = self
+      courseList.dataSource = dataSource
+      configureDataSource()
+      fetchContent()
+    }
+
+  private func configureDataSource() {
+    dataSource = UITableViewDiffableDataSource<Section, Item>(tableView: courseList) { tableView, indexPath, item -> UITableViewCell in
+      guard let cell = tableView.dequeueReusableCell(withIdentifier: CourseListCell.reuseId, for: indexPath) as? CourseListCell else { return UITableViewCell() }
+      let duration = self.timeFormatter.string(from: TimeInterval(Double(item.attributes.duration))) ?? ""
+
+      cell.courseName.text = item.attributes.name
+      cell.courseDescription.text = item.attributes.descriptionPlainText
+      cell.duration.text = "(\(duration))"
+      cell.courseType.text = item.attributes.contentType == ContentType.article ? "Article Course" : "Video Course"
+      cell.releaseDate.text = item.attributes.releasedAt.formatDateString()//.description.formatDate()
+      cell.links = item.links
+
+      if let cardArtworkUrl = URL(string: item.attributes.cardArtworkUrl) {
+        CourseAPIService.shared.fetchCardArtwork(for: cardArtworkUrl) { result in
+          switch result {
+          case .success(let artwork):
+            DispatchQueue.main.async {
+              cell.artwork.contentMode = .scaleAspectFill
+              cell.artwork.layer.cornerRadius = 5
+              cell.artwork.layer.masksToBounds = true
+              cell.artwork.image = artwork
+            }
+          case .failure(let error):
+            switch error {
+            case .imageDataError:
+              print(error.localizedDescription)
+            default:
+              print("Artwork Error: \(error.localizedDescription)")
```

Comment:
Might be a good idea to show a placeholder in this case instead of only printing an error.